book/src/dev/rfcs/0012-value-pools.md

```diff
@@ -0,0 +1,437 @@
+
+### Sapling Chain Value Pool
+
+A positive Sapling balancing value takes value from the Sapling transaction value pool and adds it to the transparent transaction value pool . A negative Sapling balancing value does the reverse. As a result, positive vbalanceSapling is treated like an input to the transparent transaction value pool , whereas negative vbalanceSapling is treated like an output from that pool.
+
+As defined in [ZIP-209], the Sapling chain value pool balance for a given block chain is the negation of the sum of all valueBalanceSapling field values for transactions in the block chain.
+
+Consensus rule: If the Sapling chain value pool balance would become negative in the block chain created as a result of accepting a block , then all nodes MUST reject the block as invalid.
+
+https://zips.z.cash/protocol/protocol.pdf#saplingbalance
+
+### Orchard Chain Value Pool
+
+Orchard introduces Action transfers, each of which can optionally perform a spend, and optionally perform an output. Similarly to Sapling, the net value of Orchard spends minus outputs in a transaction is called the Orchard balancing value, measured in zatoshi as a signed integer vbalanceOrchard.
+
+vbalanceOrchard is encoded in a transaction as the field valueBalanceOrchard. If a transaction has no Action descriptions, vbalanceOrchard is implicitly zero. Transaction fields are described in § 7.1 ‘Transaction Encoding and Consensus’ on p. 116.
+
+A positive Orchard balancing value takes value from the Orchard transaction value pool and adds it to the transpar- ent transaction value pool . A negative Orchard balancing value does the reverse. As a result, positive vbalanceOrchard is treated like an input to the transparent transaction value pool , whereas negative vbalanceOrchard is treated like an output from that pool.
+
+Similarly to the Sapling chain value pool balance defined in [ZIP-209], the Orchard chain value pool balance for a given block chain is the negation of the sum of all valueBalanceOrchard field values for transactions in the block chain.
+
+Consensus rule: If the Orchard chain value pool balance would become negative in the block chain created as a result of accepting a block , then all nodes MUST reject the block as invalid.
+
+https://zips.z.cash/protocol/protocol.pdf#orchardbalance
+
+## Proposed Implementation
+
+### Create a new `ValueBalance` type
+
+- Code will be located in a new file: `zebra-chain/src/value_balance.rs`.
+- Supported operators apply to all the `Amount`s inside the type: `+`, `-`, `+=`, `-=`, `sum()`.
+- We will use `Default` to represent a totally empty `ValueBalance`, this is the state of all pools at the genesis block.
+
+```rust
+struct ValueBalance<C> {
+    transparent: Amount<C>,
+    sprout: Amount<C>,
+    sapling: Amount<C>,
+    orchard: Amount<C>,
+}
+
+impl ValueBalance {
+    /// Consensus rule: The remaining value in the transparent transaction value pool MUST be nonnegative.
```

Comment:
@oxarbitrage where do you want to check this rule?

Reply:
I think we should do it as soon as we have the values for each pool for the transaction. 

I added this at https://github.com/ZcashFoundation/zebra/pull/2430/commits/b6bdef0c92a3d4df59aabe42f85d86134f2c9a5b initially but i made a mistake in the location so i fixed it at https://github.com/ZcashFoundation/zebra/pull/2430/commits/c9c5c741b33bb522c112af1617e4f53669000cb4